[2201.7.0] - Sync the Release 2201.7.0 branch with master (#7235)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.7.0/RELEASE_NOTE.md

    ```ballerina
    const MESSAGE = "message";
    const map<string> HTTP_OK = {httpCode: "200", message: "OK", [MESSAGE] : "BAD REQUEST"}; // Compilation error now.
    ```

- Fixed a bug in the configurable TOML syntax validation for the module structure. The error message thrown for an invalid TOML module structure is now improved to provide the variable name.

    For example, consider a non-default module `foo.bar`, which contains the following configurable variables.

    ```ballerina
    configurable int intVar = ?;
    configurable string stringVar = ?;
    ```

    If an invalid TOML structure is found for the module `foo.bar` in the `Config.toml` file, it throws an error with the following message.

    ```
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module ’foo.bar’. with variable 'intVar'. Please provide the module name as '[foo.bar]'
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module ‘foo’.bar. with variable 'stringVar'. Please provide the module name as ‘[foo.bar]’
    ```

## Platform updates

### Official support for generating GraalVM native executables

> Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see the [Build a GraalVM executable](/learn/build-a-graalvm-executable/).

Reviewer: ```suggestion
Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see [Build a GraalVM executable](/learn/build-a-graalvm-executable/).
```

Author: Fixed.

## Language updates

### New features

#### Introduction of the `group by` and `collect` clauses

> The language now supports the `group by` and `collect` clauses to perform aggregation-related operations. The `group by` clause is used to group a collection based on a `grouping-key`. The `grouping-key` will be unique for each group. The `collect` clause is used to group a collection into one group. 

Reviewer: ```suggestion
The language now supports `group by` and `collect` clauses to perform aggregation-related operations. The `group by` clause is used to group a collection based on a grouping key. The  grouping key will be unique for each group. The `collect` clause is used to group a collection into one group. 
```

Also unique to instead of unique for?

Author: Changed to `The `grouping-key` will be unique to each group.`

```ballerina

> import ballerina/io;

Reviewer: ```suggestion
import ballerina/io;

```

public type CovidEntry record {|
    string district;
    string province;
    decimal deaths;

> |};

Reviewer: ```suggestion
|};

```

public final table<CovidEntry> covidTable = table [
    {district: "Colombo", province: "Western", deaths: 21},
    {district: "Kandy", province: "Central", deaths: 14},
    {district: "Kaluthara", province: "Western", deaths: 18},
    {district: "Jaffna", province: "North", deaths: 10}

> ];

Reviewer: ```suggestion
];

```

public function main() {
    record {|string province; decimal deaths;|}[] deathsByProvince = from var {province, deaths} in covidTable
        group by province
        select {province, deaths: sum(deaths)};
    io:println(deathsByProvince);
    decimal totalDeaths = from var {deaths} in covidTable
            collect sum(deaths);

>     io:println(totalDeaths);

Reviewer: I would add the output as comments.

cc @KavinduZoysa @gimantha

Reviewer: Btw, can't we use a non-COVID example?

Author: @gimantha @KavinduZoysa

Author: As we have the COVID example used in the Learn guides also (e.g., https://ballerina.io/learn/work-with-data-using-queries-in-ballerina/#the-complete-code) will use this here and will add another example in the blog.

Reviewer 2: The Guide will be re-written again with a different examples. @KavinduZoysa Shall we change the example here?

Author: All the Get Started guides are written based on the Covid sample. Better to update them all if we are changing one and may be we can prioritize it later.

Shall we change only the aggregation example for the release note and blog for the moment?

Reviewer 3: https://github.com/ballerina-platform/ballerina-dev-website/pull/7263 Please find the new examples for group-by and collect.

```

#### Support for unions as type descriptors in constant declarations
  
Constant declarations with unions as type descriptors are now supported.

```ballerina
const float|decimal TOTAL = 1 + 2.0;
```

#### Support for more recursive type definitions

The type definition resolution logic has been improved to support more recursive type definitions.

```ballerina
type A [A];
type B [B] & readonly;
```

### Bug fixes

To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.7.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.7.0+is%3Aclosed+label%3AType%2FBug).

## Runtime updates

### Bug fixes

To view bug fixes, see the [GitHub milestone for 2201.7.0 (Swan Lake)](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+milestone%3A2201.7.0+label%3ATeam%2FjBallerina+label%3AType%2FBug+is%3Aclosed).

## Standard library updates

### New features

#### `graphql` package

- Added support for the `ID` scalar type.
- Added support for the input constraint validation.

#### `persist` package

- Added support for the `mssql` data store.

### Deprecations

#### `http` package

> - Deprecated the usage of adding interceptors in the `http:ListenerConfiguration`.

Reviewer: ```suggestion
- Deprecated the usage of adding interceptors in the `http:ListenerConfiguration` record. 
```

Don't we have to mention the alternative here? @dilanSachi @TharmiganK

Reviewer 4: Updated

Author: Fixed.

### Improvements

#### `http` package

> - Improved the default error format and message.

Reviewer: Empty line between package heading and items are inconsistent again. See difference in L179.

Author: Fixed.

---
layout: ballerina-left-nav-release-notes
title: 2201.7.0 (Swan Lake) 
permalink: /downloads/swan-lake-release-notes/2201.7.0/
active: 2201.7.0
redirect_from: 
    - /downloads/swan-lake-release-notes/2201.7.0
    - /downloads/swan-lake-release-notes/2201.7.0-swan-lake/
    - /downloads/swan-lake-release-notes/2201.7.0-swan-lake
    - /downloads/swan-lake-release-notes/
    - /downloads/swan-lake-release-notes
---

## Overview of Ballerina Swan Lake 2201.7.0

<em>2201.7.0 (Swan Lake Update 7) is the seventh update release of Ballerina Swan Lake, and it includes a new set of features and significant improvements to the compiler, runtime, standard library, and developer tooling. It is based on the 2022R4 version of the Language Specification.</em> 

## Update Ballerina

> Update your current Ballerina installation directly to 2201.7.0 by using the [Ballerina Update Tool](/learn/cli-documentation/update-tool/) as follows.

Author: Removed `by`.

                                    function () returns int idFunction,
                                    string department) 
            returns Employee => {
        name: string:'join(" ", ...name),
        // Compilation error now, since `idFunction` is not `isolated`.
        id: idFunction(),
        department
    };
    ```
  
- A bug that caused the broader type to be used instead of the singleton type when a constant is used as a type descriptor has been fixed.

  This may result in compilation errors that were previously not identified.

    ```ballerina
    const int CONST = 1 + 7;
    CONST num = 2; // Compilation error now.
    ```

  This results in a compilation error now since `num`, which is of type `CONST`, can only be integer `8`.
 
- A bug that resulted in compilation errors not being logged for duplicate keys via computed name fields of a mapping constructor in a mapping constant declaration has been fixed. This previously resulted in a panic at runtime instead.

    ```ballerina
    const MESSAGE = "message";
    const map<string> HTTP_OK = {httpCode: "200", message: "OK", [MESSAGE] : "BAD REQUEST"}; // Compilation error now.
    ```

> - Fixed a bug in the configurable TOML syntax validation for the model structure. The error message thrown for an invalid TOML module structure is now improved to provide the variable name.

Reviewer 5: ```suggestion
- Fixed a bug in the configurable TOML syntax validation for the module structure. The error message thrown for an invalid TOML module structure is now improved to provide the variable name.
```

    For example, consider a non-default module `foo.bar`, which contains the following configurable variables.

    ```ballerina
    configurable int intVar = ?;
    configurable string stringVar = ?;
    ```

    If an invalid TOML structure is found for the module `foo.bar` in the `Config.toml` file, it throws an error with the following message.

    ```
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module ’foo.bar’. with variable 'intVar'. Please provide the module name as '[foo.bar]'
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module ‘foo’.bar. with variable 'stringVar'. Please provide the module name as ‘[foo.bar]’
    ```

## Platform updates

### Official support for generating GraalVM native executables

> Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see the [Build a GraalVM executable](/learn/build-a-graalvm-executable/).

Reviewer 6: The link to the learn guide should be updated

Reviewer 6: ```suggestion
Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see the [Build a GraalVM executable](/learn/graalvm-executable-overview/).
```

Both of the following cases are allowed now.

```ballerina
const map<string> HTTP_OK = {httpCode: "200", message: "OK"};
const HTTP_BAD_REQUEST = {httpCode: "400", message: "BAD REQUEST"};
```

#### Support for unions as type descriptors in constant declarations
  
Constant declarations with unions as type descriptors are now supported.

```ballerina
const float|decimal TOTAL = 1 + 2.0;
```

#### Support for more recursive type definitions

The type definition resolution logic has been improved to support more recursive type definitions.

```ballerina
type A [A];
type B [B] & readonly;
```

### Bug fixes

To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.7.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.7.0+is%3Aclosed+label%3AType%2FBug).

## Runtime updates

### Bug fixes

To view bug fixes, see the [GitHub milestone for 2201.7.0 (Swan Lake)](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+milestone%3A2201.7.0+label%3ATeam%2FjBallerina+label%3AType%2FBug+is%3Aclosed).

## Standard library updates

### New features

#### `http` package

> - Introduced the `httpscerr` module, which contains the HTTP status code errors.

Reviewer 7: Do we need to include this to release note?

Reviewer 6: No, will remove it.

Reviewer 6: Removed via [89a9dda](https://github.com/ballerina-platform/ballerina-dev-website/pull/7235/commits/89a9dda02f126b50dcc8ea8a5774b7f215bd7905)

## Developer tools updates

### New features

#### Language Server

- Added a new API named `workspaceProjects` to the workspace manager to get the projects loaded in the workspace.

#### OpenAPI Tool
- Added support to generate low-level service skeletons without any data-binding logic. This mode can be enabled by adding the `--without-data-binding` flag to the OpenAPI to Ballerina service generation command. This new generation mode can be useful for generating pass-through and proxy services based on a provided OpenAPI contract.

### Improvements

#### Language Server

- Improved the order of completion items in the client resource access action context.
- Improved the documentation generated by the `Document this` code action to follow the Ballerina code-style best practices.
- Improved the placeholder support provided for the completion items of the service-template snippets.

#### CLI commands

- Renamed the `--native` flag to `--graalvm`.
- Introduced the `--graalvm-build-options` build option to support passing additional arguments for GraalVM native image generation.

#### OpenAPI tool

> - Added auto-generated file headers for all the generated Ballerina files in OpenAPI to Ballerina service and client generations. Also, the users will have the option to  replace the auto generated headers with their own license, using the `--license <license-file-path>` command option.

Reviewer 7: ```suggestion
- Added auto-generated file headers for all the generated Ballerina files in OpenAPI to Ballerina service and client generations. Also, the users will have the option to replace the auto-generated headers with their own license using the `--license <license-file-path>` command option.
```

title: 2201.7.0 (Swan Lake) 
permalink: /downloads/swan-lake-release-notes/2201.7.0/
active: 2201.7.0
redirect_from: 
    - /downloads/swan-lake-release-notes/2201.7.0
    - /downloads/swan-lake-release-notes/2201.7.0-swan-lake/
    - /downloads/swan-lake-release-notes/2201.7.0-swan-lake
    - /downloads/swan-lake-release-notes/
    - /downloads/swan-lake-release-notes
---

## Overview of Ballerina Swan Lake 2201.7.0

<em>2201.7.0 (Swan Lake Update 7) is the seventh update release of Ballerina Swan Lake, and it includes a new set of features and significant improvements to the compiler, runtime, standard library, and developer tooling. It is based on the 2023R1 version of the Language Specification.</em> 

## Update Ballerina

Update your current Ballerina installation directly to 2201.7.0 using the [Ballerina Update Tool](/learn/cli-documentation/update-tool/) as follows.

1. Run `bal update` to get the latest version of the Update Tool.
2. Run `bal dist update` to update to this latest distribution.

## Install Ballerina

If you have not installed Ballerina, download the [installers](/downloads/#swanlake) to install.

## Backward-incompatible changes

> - A bug that allowed using a function reference of a `non-isolated` function type in a function or method call expression within an `isolated` function or method has been fixed.

Reviewer: ```suggestion
- A bug that allowed using a function reference of a non-`isolated` function type in a function or method call expression within an `isolated` function or method has been fixed.
```

Same comment re: making changes to already approved content - https://github.com/ballerina-platform/ballerina-dev-website/pull/7156/files#diff-a2d9e6006fef3a642196e95b059536ee565b283e4dbdeaa8b22bd3892b4104c9R31.

    ```ballerina
    type Employee record {|
        string name;
        int id;
        string department;
    |};

    isolated function createEmployee(string[] name,
            function () returns int idFunction,
            string department)
            returns Employee => {
        name: string:'join(" ", ...name),
        // Compilation error now, since `idFunction` is not `isolated`.
        id: idFunction(),
        department
    };  
    ```
  
- A bug that caused the broader type to be used instead of the singleton type when a constant is used as a type descriptor has been fixed.

>   For example, the code below may result in a compilation error now since `num`, which is of type `CONST` can only be integer `8`.

Author: Reverted.

>   For example, the code below may result in a compilation error now since `num`, which is of type `CONST` can only be integer `8`.

Reviewer: Still seems to be different.

>   For example, the code below may result in a compilation error now since `num`, which is of type `CONST` can only be integer `8`.

Author: Fixed.

    ```ballerina
    const int CONST = 1 + 7;
    CONST num = 2; // Compilation error now.
    ```

- A bug that resulted in compilation errors not being logged for duplicate keys via computed name fields of a mapping constructor in a mapping constant declaration has been fixed. This previously resulted in a panic at runtime instead.

    ```ballerina
    const MESSAGE = "message";
    const map<string> HTTP_OK = {httpCode: "200", message: "OK", [MESSAGE] : "BAD REQUEST"}; // Compilation error now.
    ```

> - Fixed a bug in the configurable TOML syntax validation for the module structure. The error message thrown for an invalid TOML module structure is now improved to provide the variable name.

Reviewer: ```suggestion
- Fixed a bug in the configurable TOML syntax validation for the module structure. The error message for an invalid TOML module structure is now improved to provide the variable name.
```

> - Fixed a bug in the configurable TOML syntax validation for the module structure. The error message thrown for an invalid TOML module structure is now improved to provide the variable name.

Reviewer: How is this a bug/breaking change?

Author: @warunalakshitha @HindujaB @Nadeeshan96

Reviewer 8: Error message changed. So any tests which will depend on previous error message will break.

    For example, consider a non-default module named `foo.bar`, which contains the following configurable variables.

    ```ballerina
    configurable int intVar = ?;
    configurable string stringVar = ?;
    ```

>     If an invalid TOML structure is found for the `foo.bar` module in the `Config.toml` file, it throws an error with the following message.

Reviewer: ```suggestion
    If an invalid TOML structure is found for the `foo.bar` module in the `Config.toml` file, it results in the following error messages.
```

    ```

>     [Config.toml:(1:1,2:23)] invalid TOML structure found for module ’foo.bar’. with variable 'intVar'. Please provide the module name as '[foo.bar]'

Reviewer: ```suggestion
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar'. with variable 'intVar'. Please provide the module name as '[foo.bar]'
```

>     [Config.toml:(1:1,2:23)] invalid TOML structure found for module ‘foo’.bar. with variable 'stringVar'. Please provide the module name as ‘[foo.bar]’

Reviewer: Backticks need to be fixed.

Author: @HindujaB @Nadeeshan96 @warunalakshitha

Reviewer 5: ```suggestion
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar' with variable 'stringVar'. Please provide the module name as '[foo.bar]'
```

    ```

> ## Platform updates

Reviewer: Is this the first section intentionally?

Author: Yes. Is there a reason to not make it the first section? @MaryamZi

Reviewer: Because we usually have lang updates first, then stdlib, etc.

Author: Added the topic at the top because to highlight it as it is one of the main features in this release and also because it affects the entire platform in general.

### Official support for generating GraalVM native executables

Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see [Build a GraalVM executable](/learn/graalvm-executable-overview/).

## Language updates

### New features

#### Introduction of the `group by` and `collect` clauses

> The language now supports the `group by` and `collect` clauses to perform aggregation-related operations. The `group by` clause is used to group a collection based on a `grouping-key`. The `grouping-key` will be unique to each group. The `collect` clause is used to group a collection into one group. 

Reviewer: We emphasize only keywords, construct names, etc. So we shouldn't emphasize `grouping-key`.

Author: Updated.

```ballerina
import ballerina/io;

type Order record {|
    string name;
    int price;
|};

public function main() {
    Order[] orders = [{name: "Item1", price: 91}, {name: "Item2", price: 83}, {name: "Item1", price: 75}, {name: "Item2", price: 88}];
    var averages = from var {name, price} in orders
                        group by name
                        select {name, avg: avg(price)}; 
    io:println(averages); // [{"name":"Item1","avg":83},{"name":"Item2","avg":85.5}]
}
```

> ```

Reviewer: ```suggestion
```
````suggestion
```ballerina
````

> ```

Reviewer: Can we move the last line "The `collect` clause is used to group a collection into one group." from L95 to immediately before this sample?

Author: The line is added before the sample under the respective sub-topic.

    ```

## Platform updates

### Official support for generating GraalVM native executables

Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see [Build a GraalVM executable](/learn/graalvm-executable-overview/).

## Language updates

### New features

#### Introduction of the `group by` and `collect` clauses

The language now supports the `group by` and `collect` clauses to perform aggregation-related operations.

The `group by` clause is used to group a collection based on a grouping-key as shown in the example below. The grouping-key will be unique to each group.

```ballerina
import ballerina/io;

type Order record {|
    string name;
    int price;
|};

public function main() {
    Order[] orders = [{name: "Item1", price: 91}, {name: "Item2", price: 83}, {name: "Item1", price: 75}, {name: "Item2", price: 88}];
    var averages = from var {name, price} in orders
        group by name

>         select {name, avg: avg(price)};

Reviewer: Weren't we printing the result here?

Author: @KavinduZoysa

}
```

> The `collect` clause is used to group a collection into one group as shown in the example below.

Reviewer: Is " as shown in the example below" really required?

```suggestion
The `collect` clause is used to group a collection into one group.
```

Author: There is no harm adding it to connect the description with the code below.

    [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar' with variable 'stringVar'. Please provide the module name as '[foo.bar]'
    ```

## Platform updates

### Official support for generating GraalVM native executables

Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see [Build a GraalVM executable](/learn/graalvm-executable-overview/).

## Language updates

### New features

#### Introduction of the `group by` and `collect` clauses

The language now supports the `group by` and `collect` clauses to perform aggregation-related operations.

The `group by` clause is used to group a collection based on a grouping-key. The grouping-key will be unique to each group.

```ballerina
import ballerina/io;

type Order record {|
    string name;
    int price;
|};

public function main() {

>     Order[] orders = [{name: "Item1", price: 91}, {name: "Item2", price: 83}, {name: "Item1", price: 75}, {name: "Item2", price: 88}];

Reviewer: ```suggestion
    Order[] orders = [
        {name: "Item1", price: 91},
        {name: "Item2", price: 83},
        {name: "Item1", price: 75},
        {name: "Item2", price: 88}
    ];
```

Reviewer: I would use real-world item names over Item1, Item2, etc. cc @KavinduZoysa

Author: Updated.

    var averages = from var {name, price} in orders
        group by name
        select {name, avg: avg(price)};
    io:println(averages); // [{"name":"Item1","avg":83},{"name":"Item2","avg":85.5}]
}
```

The `collect` clause is used to group a collection into one group.

```ballerina
import ballerina/io;

type Order record {|
    string name;
    int price;
|};

public function main() {

>     Order[] orders = [{name: "Item1", price: 91}, {name: "Item2", price: 83}, {name: "Item1", price: 75}, {name: "Item2", price: 88}];

Reviewer: ```suggestion
    Order[] orders = [
        {name: "Item1", price: 91},
        {name: "Item2", price: 83},
        {name: "Item1", price: 75},
        {name: "Item2", price: 88}
    ];
```

  For example, the code below may result in a compilation error now since `num`, which is of type `CONST` can only be integer `8`.

    ```ballerina
    const int CONST = 1 + 7;
    CONST num = 2; // Compilation error now.
    ```

- A bug that resulted in compilation errors not being logged for duplicate keys via computed name fields of a mapping constructor in a mapping constant declaration has been fixed. This previously resulted in a panic at runtime instead.

    ```ballerina
    const MESSAGE = "message";
    const map<string> HTTP_OK = {httpCode: "200", message: "OK", [MESSAGE] : "BAD REQUEST"}; // Compilation error now.
    ```

- Fixed a bug in the configurable TOML syntax validation for the module structure. The error message for an invalid TOML module structure is now improved to provide the variable name.

    For example, consider a non-default module named `foo.bar`, which contains the following configurable variables.

    ```ballerina
    configurable int intVar = ?;
    configurable string stringVar = ?;
    ```

    If an invalid TOML structure is found for the `foo.bar` module in the `Config.toml` file, it results in the following error messages.

    ```

>     [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar'. with variable 'intVar'. Please provide the module name as '[foo.bar]'

Reviewer 5: ```suggestion
    [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar' with variable 'intVar'. Please provide the module name as '[foo.bar]'
```

    [Config.toml:(1:1,2:23)] invalid TOML structure found for module 'foo.bar' with variable 'stringVar'. Please provide the module name as '[foo.bar]'
    ```

## Platform updates

### Official support for generating GraalVM native executables

Ballerina now officially supports generating GraalVM native executables and the language and the standard libraries are compatible with the GraalVM native executable generation. In addition to that, GraalVM incompatibility warnings will be printed for any incompatible modules in the application. To explore more on this support, see [Build a GraalVM executable](/learn/graalvm-executable-overview/).

## Language updates

### New features

#### Introduction of the `group by` and `collect` clauses

The language now supports the `group by` and `collect` clauses to perform aggregation-related operations.

> ##### The `group by` clause

Reviewer: IMO, these subheadings aren't really necessary.

Author: Removed.

> The `group by` clause is used to group a collection based on a `grouping-key` as shown in the example below. The `grouping-key` will be unique to each group.

Reviewer: ```suggestion
The `group by` clause is used to group a collection based on a grouping-key as shown in the example below. The grouping-key will be unique to each group.
```

Author: Fixed.

```ballerina
import ballerina/io;

type Order record {|
    string name;
    int price;
|};

public function main() {
    Order[] orders = [{name: "Item1", price: 91}, {name: "Item2", price: 83}, {name: "Item1", price: 75}, {name: "Item2", price: 88}];
    var averages = from var {name, price} in orders
        group by name
        select {name, avg: avg(price)};
```ballerina

> }

Reviewer: Invalid. Please recheck and fix changes.

Author: Fixed.

    io:println(average); // 19.375
}
```

### Improvements

#### Support for constant declarations without a type descriptor

Constants can now be declared without specifying a type descriptor. Previously, it was required to specify a type descriptor except with numeric and string literal expressions.

Both of the following cases are allowed now.

```ballerina
const map<string> HTTP_OK = {httpCode: "200", message: "OK"};
const HTTP_BAD_REQUEST = {httpCode: "400", message: "BAD REQUEST"};
```

#### Support for unions as type descriptors in constant declarations
  
Constant declarations with unions as type descriptors are now supported.

```ballerina
const float|decimal TOTAL = 1 + 2.0;
```

#### Support for more recursive type definitions

The type definition resolution logic has been improved to support more recursive type definitions.

```ballerina
type A [A];
type B [B] & readonly;
```

### Bug fixes

To view bug fixes, see the [GitHub milestone for Swan Lake 2201.7.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.7.0+is%3Aclosed+label%3AType%2FBug).

## Runtime updates

### Bug fixes

> To view bug fixes, see the [GitHub milestone for 2201.7.0 (Swan Lake)](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+milestone%3A2201.7.0+label%3ATeam%2FjBallerina+label%3AType%2FBug+is%3Aclosed).

Reviewer 7: Why using two different ways for link text. Compare this and line 174.

Author: Fixed.
